compiler/rustc_session/src/session.rs

```diff
@@ -586,10 +586,15 @@ impl Session {
     }
 
     pub fn mir_opt_level(&self) -> usize {
-        self.opts
-            .unstable_opts
-            .mir_opt_level
-            .unwrap_or_else(|| if self.opts.optimize != OptLevel::No { 2 } else { 1 })
+        self.opts.unstable_opts.mir_opt_level.unwrap_or_else(|| {
+            if self.target.has_convergent {
+                // Special care needs to be taken for convergent operations (i.e. not duplicating them).
+                // We do not want to handle these, so do not run any optimizations.
+                0
```

Comment:
Someone shouldn't be able to override the MIR opt level to a higher one, enabling unsound optimizations.